packages/engineer/feature/dev-server.dev-server.env.ts

```diff
@@ -0,0 +1,282 @@
+    ) => {
+        const application = new ApplicationProxyService({ basePath });
+        const overrideConfigsMap = new Map<string, OverrideConfig>();
+        const disposables = new Set<() => unknown>();
+        let nodeEnvManager: NodeEnvironmentsManager | null = null;
+
+        const getNodeEnvManager = () => nodeEnvManager;
+
+        // Extract these into a service
+        const runFeature = async ({
+            featureName,
+            runtimeOptions = {},
+            configName,
+            overrideConfig,
+        }: IRunFeatureOptions) => {
+            if (overrideConfig) {
+                const generatedConfigName = generateConfigName(configName);
+                overrideConfigsMap.set(generatedConfigName, {
+                    overrideConfig: Array.isArray(overrideConfig) ? overrideConfig : [],
+                    configName,
+                });
+                configName = generatedConfigName;
+            }
+            // clearing because if running features one after the other on same engine, it is possible that some measuring were done on disposal of stuff, and the measures object will not be re-evaluated, so cleaning it
+            performance.clearMeasures();
+            performance.clearMarks();
+            return getNodeEnvManager()!.runServerEnvironments({
+                featureName,
+                configName,
+                overrideConfigsMap,
+                runtimeOptions,
+                mode: nodeEnvironmentsMode,
+            });
+        };
+
+        const closeFeature = ({ featureName, configName }: IFeatureMessagePayload) => {
+            if (configName) {
+                overrideConfigsMap.delete(configName);
+            }
+            performance.clearMeasures();
+            performance.clearMarks();
+            return getNodeEnvManager()!.closeEnvironment({
+                featureName,
+                configName,
+            });
+        };
+        const getMetrics = () => {
+            return {
+                marks: performance.getEntriesByType('mark'),
+                measures: performance.getEntriesByType('measure'),
+            };
+        };
+
+        const close = async () => {
+            for (const dispose of disposables) {
+                await dispose();
+            }
+            disposables.clear();
+        };
+
+        run(async () => {
+            // Should engine config be part of the dev experience of the engine????
+            const engineConfig = await application.getEngineConfig();
+            if (engineConfig && engineConfig.require) {
+                await application.importModules(engineConfig.require);
+            }
+
+            const { port, app, close, socketServer } = await launchHttpServer({
+                staticDirPath: application.outputPath,
+                httpServerPort,
+            });
+            disposables.add(() => close());
+
+            const host = new WsServerHost(socketServer.of(`/${devServerEnv.env}`));
```

Comment:
The next 4 lines should be considered to be wrapped as a method in communication